### PR TITLE
6914 CHT Script API

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

This PR:
- Adds a new cht api for targets, tasks and contact summary with 2 sync functions: hasPermissions, hasAnyPermission.
- Refactors Admin app and Webapp to use the api functions.
 
https://github.com/medic/cht-core/issues/6914

CHT-Conf: https://github.com/medic/medic-conf/pull/409
CHT-Docs: https://github.com/medic/cht-docs/pull/518 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
Hi @garethbowen I'm still waiting for the solution feedback from the app services folks, they've been busy. However, I was wondering if you can please give early feedback? Thanks for your time!

File: webapp/src/ts/services/contact-summary.service.ts
Comment (kennsippell) [inline webapp/src/ts/services/contact-summary.service.ts:105]: this approach is difficult to write automated tests for partner code. how do you recommend people to write unit tests for code which references this api? how can we architect for testability?

Comment (garethbowen) [reply webapp/src/ts/services/contact-summary.service.ts:105]: Can you elaborate on why this would be difficult to test? Firstly do you mean unit tests, integration tests, or both?

I would expect this to be unit testable by mocking out the value for `cht` in the script to do whatever you want. We could develop a stub for this if that would be easier than mocking out the entire API.

For full integration tests then it would use the real logged in user.

For partial integration tests it depends where you're drawing the line.

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/cht-script-api.service.ts:95]: This code appears to be copied from the Auth service or similar. We need to reuse this rather than copying it because it's easy to get it wrong. This could be as simple as one of these services calling the other, or this could be extracted into a shared lib which might then also reduce the duplication in api, admin, sentinel, etc.

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/cht-script-api.service.ts:100]: I'm hoping `hasRole` isn't needed. Roles are usually an implementation detail for working out if a user has permission to do something. Everything should be permission based. Remove this unless it can be proven that permissions aren't enough.

File: webapp/src/ts/services/contact-summary.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/contact-summary.service.ts:105]: Everywhere we expose the API it needs to be versioned, eg:

```suggestion
      const summary = generatorFunction(contact, reports || [], lineage || [], uhcStats, { v1: chtScriptApi }, targetDoc);
```

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts:95]: Yeah, it's a simplified sync version of the one in Auth service, without checking for disallowed permissions (as Auth service do) which I haven't found usages of it in webapp  🤔

I thought about that and actually tried to use the one in Auth service but it needs to be simplified and refactored so we have the async function and sync function that uses the same checking function 😁 If that okay I'll go and refactor Auth Service code.

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts:100]: Okay, I'll ask if this is needed for the current ticket once I get feedback

File: webapp/src/ts/services/contact-summary.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: The service is already returning the api in a version obj, I rather have the versions managed in 1 place than in multiple places 🤔

Comment (garethbowen) [reply webapp/src/ts/services/contact-summary.service.ts:105]: Ahh ok! Then the name is wrong. Rename it to just `this.chtScriptApiService.get();` because one day it'll return `{ v1: {}, v2: {} }`

Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: oh right! [facepalm] I'm going to rename it

Comment (kennsippell) [reply webapp/src/ts/services/contact-summary.service.ts:105]: how can we support this in https://github.com/medic/medic-conf-test-harness? specifically, passing angular services into partner code makes that integration test pattern difficult. is it possible to implement the `cht` api in cht-conf so it is compiled in the app bundle? to pass in an npm module (like Utils which already has access to the app_settings) or to abstract this into a shared-lib for re-use in integraiton tests?

Comment (garethbowen) [reply webapp/src/ts/services/contact-summary.service.ts:105]: I like the shared lib idea, because there are parts of this we would want to use in api, sentinel, etc too. Ultimately this should be mocked because the configuration test will be opinionated about the user's roles or permissions in order to assert the correct outcomes. @latin-panda Can you attempt to write or modify a test using the test harness to assert that this design is testable?

The problem with implementing any of this in medic-conf or an npm library like Utils is we want the API to be implementation agnostic, that is to say, if Product decides to change the way permissions are done we want the API to keep working as much as possible. If we include this API in CHT Core then we can guarantee the versions are in sync, but otherwise you get complicated dependencies between Utils, medic-conf, and CHT Core.

> are you already planning to refactor the new uhcStats object into the api prior to 3.12 release?

Unfortunately that work is already ATed, merged, etc so at the moment we'll leave it as is. Once the API design is sorted and we're happy with it (hopefully before 3.12 is released) then we should revisit the uhcStats solution to see if this API is an improvement.

Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: Looking at it :)

Comment (kennsippell) [reply webapp/src/ts/services/contact-summary.service.ts:105]: @latin-panda Ping me with any questions about medic-conf-test-harness. Happy to give you [pointers](https://github.com/medic/medic-conf-test-harness/blob/master/all-chts-bundle.js#L28) or fill in your [understanding](https://github.com/medic/medic-conf-test-harness/pull/102#issue-644505398).

Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: I think it should be possible adding the cht-api in the test medic-conf-test-harness, but I'm still struggling to run my test in the default config.
`npm ERR! errno ENOTFOUND
npm ERR! network request to https://registry.npmjs.org/cht-core-3-10 failed, reason: getaddrinfo ENOTFOUND registry.npmjs.org
npm ERR! network This is a problem related to network connectivity.
`
I think it's going to be something like this as Kenn suggested: 
Having the cht-api in the shared-libs, but it will need to receive the db documents to do the checks and being able to keep the functions sync for tasks, targets, etc. 

Add the reference:
Screen Shot 2021-06-10 at 7 13 50 pm

Add the mock data, for example:
Screen Shot 2021-06-10 at 7 15 00 pm

Pass the api: 
Screen Shot 2021-06-10 at 7 14 44 pm

I'll continue with this implementation in cht-core about having the shared-libs tomorrow :)

Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: I was able to mock the api in shared-libs and run the medic-conf-test-harness for tasks successfully :) 

The example in the screenshots are about an user with a custom permission "can_make_delivery_form" and another user without that permission. 

Screen Shot 2021-06-11 at 2 03 01 pm

Screen Shot 2021-06-11 at 2 05 55 pm

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts:100]: I've removed `hasRole`

File: webapp/src/ts/services/contact-summary.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/contact-summary.service.ts:105]: Fixed

File: shared-libs/cht-script-api/src/index.js
Comment (latin-panda) [inline shared-libs/cht-script-api/src/index.js:58]: The `setChtCoreSettingsDoc` and the `setUserSettingsDoc` aren't exposed in the `v1 api`, because these aren't meant for regular users as these affects the final result the `hasPermission`.

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts]: Done

**Comment by latin-panda** [comment]
Hi @garethbowen I've refactored the Auth service from webapp and admin app to use this api, can you please provide feedback? Thanks!

**Comment by garethbowen** [comment]
Sorry @latin-panda I'm flat out this week. @dianabarsan Would you mind having a look? Otherwise I'll make sure to get to this next week.

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [inline shared-libs/cht-script-api/src/auth.js:6]: @dianabarsan should we check for `admin` as well? 
The admin document has the role without underscore:
Screen Shot 2021-06-17 at 1 35 13 pm
While the cookie has both: `{"name":"admin","roles":["_admin","admin"],"home":"/admin/"}`

File: shared-libs/cht-script-api/test/auth.spec.js
Comment (dianabarsan) [inline shared-libs/cht-script-api/test/auth.spec.js:112]: The auth service spec has a few more test cases which have value to be added here, along with: 
- a test case with multiple permissions and multiple roles
- a test case with multiple permissions spread across roles
- a test case with an unknown disallowed permission
- a test case with no roles
- a test case with an empty permission (empty string) 
-

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (dianabarsan) [inline webapp/src/ts/services/cht-script-api.service.ts:29]: I don't think this is ever destroyed.

File: webapp/tests/karma/ts/services/contact-summary.service.spec.ts
Comment (dianabarsan) [inline webapp/tests/karma/ts/services/contact-summary.service.spec.ts:29]: Please prefer to use `sinon.stub()` instead of noop functions.

File: webapp/tests/karma/ts/services/rules-engine.service.spec.ts
Comment (dianabarsan) [inline webapp/tests/karma/ts/services/rules-engine.service.spec.ts:78]: Please use `sinon.stub()` instead of noop functions.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! I left a couple of requests inline.

File: shared-libs/cht-script-api/src/index.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/index.js:79]: I've been wanting to remove these functions and cache thing from this script - I'm going to try rewire 🤞

File: shared-libs/cht-script-api/README.md
Comment (latin-panda) [reply shared-libs/cht-script-api/README.md:13]: The idea is to provide more helper functions about different topics (authorization, uhc, etc) and at some point minimise/remove the other globals in tasks, targets and contact-summary. This can be used in `/api`, `/admin` and `/webapp` internally as well. 
I think we're trying out first by providing a way to check for permissions, people can then access them through a `cht` namespace: `const myVar = cht.v1.myFunction();`

(@garethbowen pls correct me as need)

Comment (latin-panda) [reply shared-libs/cht-script-api/README.md:13]: is this folder name okay? What do you think? I wanted to differentiate from our `/api` app but naming it `cht-api` is still unclear (I guess 🤔)

What do you think about having the auth code for checking permissions as a section in /cht-script-api ? I mean `/cht-script-api/auth.js` 
it could also be its own thing in `shared-libs/auth` but since it's accessed through this new api, I've put it inside to motivate the access through the new api.

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:11]: Nice!

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:38]: Sure!

File: shared-libs/cht-script-api/test/auth.spec.js
Comment (latin-panda) [reply shared-libs/cht-script-api/test/auth.spec.js:112]: Nice! I'm adding these

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:6]: Thanks for the info! Instead of using the user-settings directly, I switched it to use the userCtx from session service 👍

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:92]: Oh right 😵‍💫 Fixed!

File: webapp/tests/karma/ts/services/contact-summary.service.spec.ts
Comment (latin-panda) [reply webapp/tests/karma/ts/services/contact-summary.service.spec.ts:29]: Fixed!

File: webapp/tests/karma/ts/services/rules-engine.service.spec.ts
Comment (latin-panda) [reply webapp/tests/karma/ts/services/rules-engine.service.spec.ts:78]: Fixed!

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts:29]: Yeah, since this service is injected in the root, so not much benefit from unsubscribing observables. 
I'm wondering about the tests like e2e, it should be fine 🤔 but do you think this could be a memory leak problem for e2e?

Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts]: Removed, should be fine

File: shared-libs/cht-script-api/src/index.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/index.js]: This code was removed

File: shared-libs/cht-script-api/src/index.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/index.js]: This code was removed

File: shared-libs/cht-script-api/README.md
Comment (latin-panda) [reply shared-libs/cht-script-api/README.md:13]: I've removed getApi

File: shared-libs/cht-script-api/src/index.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/index.js]: I've removed the functions and cache from this library, found a way for tests harness by wrapping this library to pass the user and settings from the test case configuration.
Webapp will handle this library to keep the functions (hasPermissions, hasAnyPermission) synchronous for tasks, targets and contact-summary when user doc and cht setting aren't passed as parameters.

**Comment by latin-panda** [comment]
Hi @dianabarsan, can you please review again? Thanks!

File: shared-libs/cht-script-api/test/auth.spec.js
Comment (latin-panda) [reply shared-libs/cht-script-api/test/auth.spec.js:162]: Adding -> test case with multiple permissions and multiple roles

Comment (latin-panda) [reply shared-libs/cht-script-api/test/auth.spec.js:181]: Unit test added :)

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work! I think this is very close to be ready, I just found a couple of instances where calls to one api endpoint with incorrect params will throw errors wouldn't be easy to understand for app developers.

File: webapp/tests/karma/ts/services/cht-script-api.service.spec.ts
Comment (latin-panda) [reply webapp/tests/karma/ts/services/cht-script-api.service.spec.ts:273]: 🤔 I'm going to print a debug message for these 2 scenarios from the CHT Script API library

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:33]: Done!

File: shared-libs/cht-script-api/src/auth.js
Comment (latin-panda) [reply shared-libs/cht-script-api/src/auth.js:56]: Done!

File: shared-libs/cht-script-api/test/index.spec.js
Comment (latin-panda) [reply shared-libs/cht-script-api/test/index.spec.js:25]: Done! I have stub the functions from auth, checked parameters being passed.

File: webapp/tests/karma/ts/services/cht-script-api.service.spec.ts
Comment (latin-panda) [reply webapp/tests/karma/ts/services/cht-script-api.service.spec.ts:273]: Done, I've added checks for these in the unit test of auth.spec.js

File: webapp/src/ts/services/cht-script-api.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/cht-script-api.service.ts:51]: Done

**Comment by latin-panda** [comment]
Hi @dianabarsan can you please have a look again? Thanks for your time and feedback?

File: shared-libs/cht-script-api/test/index.spec.js
Comment (dianabarsan) [inline shared-libs/cht-script-api/test/index.spec.js:54]: Please use `to.deep.equal` here. `have.members` ignores order, which is pretty important in the case of arguments passed.

File: shared-libs/cht-script-api/test/index.spec.js
Comment (dianabarsan) [inline shared-libs/cht-script-api/test/index.spec.js:31]: Please use `to.deep.equal` here as well.

File: webapp/src/ts/services/rules-engine.service.ts
Comment (dianabarsan) [inline webapp/src/ts/services/rules-engine.service.ts:109]: The arguments passed to `getRulesSettings` seem to be identical to the ones passed to `monitorChanges` and `rulesConfigChange`. Do you think it would be an idea to create a function that generates an object with this properties and have just one argument for each of these functions?

**Review by dianabarsan — APPROVED** [review]
Awesome! This is a great addition! 

I left two minor requests related to testing and one refactoring suggestion which is optional, but approving so you're not blocked.

File: shared-libs/cht-script-api/test/index.spec.js
Comment (latin-panda) [reply shared-libs/cht-script-api/test/index.spec.js]: Done!

File: shared-libs/cht-script-api/test/index.spec.js
Comment (latin-panda) [reply shared-libs/cht-script-api/test/index.spec.js]: Done!

File: webapp/src/ts/services/rules-engine.service.ts
Comment (latin-panda) [reply webapp/src/ts/services/rules-engine.service.ts]: Done!

**Pull request merged**
